packages/nimble-components/src/wafer-map/modules/experimental/worker-renderer.ts

```diff
@@ -1,14 +1,65 @@
+import { transfer } from 'comlink';
 import type { WaferMap } from '../..';
 import { HoverDieOpacity } from '../../types';
 
 /**
  * Responsible for drawing the dies inside the wafer map, adding dieText and scaling the canvas
  */
 export class WorkerRenderer {
+    private isWorkerAlive = false;
+
     public constructor(private readonly wafermap: WaferMap) {}
 
-    public drawWafer(): void {
-        // rendering will be implemented in a future PR
+    public async setupAndDrawWafer(): Promise<void> {
+        if (this.wafermap.diesTable === undefined) {
+            return;
+        }
+        await this.setupWorker();
+        const columnIndexes = Int32Array.from(
+            this.wafermap.diesTable.getChild('colIndex')!.toArray()
+        );
+        const rowIndexes = Int32Array.from(
+            this.wafermap.diesTable.getChild('rowIndex')!.toArray()
+        );
+        await this.wafermap.worker.setColumnIndexes(
+            transfer(columnIndexes, [columnIndexes.buffer])
+        );
+        await this.wafermap.worker.setRowIndexes(
+            transfer(rowIndexes, [rowIndexes.buffer])
+        );
+        await this.drawWafer();
+    }
+
+    public async drawWafer(): Promise<void> {
+        if (this.wafermap.diesTable === undefined) {
+            return;
+        }
+
+        if (this.wafermap.worker === undefined) {
+            throw new Error('Worker is not defined');
```

Comment:
It's pretty rare that we throw exceptions in Nimble because we don't ever want to leave clients components in a bad state. I just want to double check that a client hitting this would indicate a programming error and not a case we think anyone would likely hit.

Reply:
I thought about this some more and I think I'd advocate for removing the check for `undefined` and the `throw`. Instead you can use the [non-null assertion operator](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) (`!`) to tell the compiler that you know the value can never be undefined. Even if that assertion turns out to be incorrect, at runtime JS will throw its own error when something tries to use the undefined value so you'll still get equivalent runtime behavior as you have now, but with less code.

If you want to avoid using `!` everywhere, you could do `const worker = this.wafermap.worker!;` once at the beginning of the method and then use the `worker` variable throughout.

Reply:
A slightly bigger refactor to consider would be for this renderer to own the worker rather than the wafer map. This class is the one that does most of the work on the worker so it's the one that cares about its lifecycle the most. Why should it have to keep asking the component for something that it could own itself?

From a quick scan I think this would be a better direction, but I'll leave it up to you and @munteannatan to evaluate if it is feasible.

Reply:
I deleted the null-check and throwing and refactored the code so worker-renderer owns the worker, the changes can be seen here https://github.com/ni/nimble/pull/1929/commits/e9d99ae9ae3ccc5d1972be552efaceb0b0c919ee

Reply:
Thanks, I like how that turned out.